mace/tools/slurm_distributed.py

```diff
@@ -0,0 +1,27 @@
+###########################################################################################
+# Slurm environment setup for distributed training.
+# This code is refactored from rsarm's contribution at:
+# https://github.com/Lumi-supercomputer/lumi-reframe-tests/blob/main/checks/apps/deeplearning/pytorch/src/pt_distr_env.py
+# This program is distributed under the MIT License (see MIT.md)
+###########################################################################################
+
+import os
+import hostlist
+
+
+class DistributedEnvironment():
+    def __init__(self):
+        self._setup_distr_env()
+        self.master_addr = os.environ['MASTER_ADDR']
+        self.master_port = os.environ['MASTER_PORT']
+        self.world_size = int(os.environ['WORLD_SIZE'])
+        self.rank = int(os.environ['RANK'])
+        self.local_rank = int(os.environ['LOCAL_RANK'])
+
+    def _setup_distr_env(self):
+        hostnames = hostlist.expand_hostlist(os.environ['SLURM_JOB_NODELIST'])
+        os.environ['MASTER_ADDR'] = hostnames[0]
+        os.environ['MASTER_PORT'] = '33333' # arbitrary
```

Comment:
It's unlikely that port 33333 will be occupied, but it might be better to use the following to guarantee that a open port is used:
```
from socket import socket

s = socket()
s.bind((hostnames[0], 0)
os.environ['MASTER_PORT'] = s.getsockname()[1]
```

Reply:
This is much cleaner on paper, I agree -- but it's not quite applicable here, because run_train.py is launched multiple times, once for each training process, so on each process, socket() will find a different port.

Now, we could use socket() to choose the port on only one of the processes, but then we have to find some means to communicate the port to the other processes before the distributed communication has been initialized, which gets messy. It might be preferable to simply allow the port to be set in the shell with an environment variable:

`def _setup_distr_env(self):
        hostname = hostlist.expand_hostlist(os.environ['SLURM_JOB_NODELIST'])[0]
        os.environ['MASTER_ADDR'] = hostname
        os.environ['MASTER_PORT'] = os.environ.get('MASTER_PORT', '33333')`